pallets/fiat_filter/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+            }
+            Some(revoke::Call::revoke(_revoke, _proof)) => return Ok(Permill::from_percent(81)),
+            Some(revoke::Call::unrevoke(_unrevoke, _proof)) => {
+                return Ok(Permill::from_percent(82))
+            }
+            Some(revoke::Call::remove_registry(_rm, _proof)) => {
+                return Ok(Permill::from_percent(83))
+            }
+            _ => {}
+        };
+        match call.is_sub_type() {
+            Some(attest::Call::set_claim(_attester, _att, _sig)) => {
+                return Ok(Permill::from_percent(40))
+            }
+            _ => {}
+        }
+
+        // return error if Call not in list
+        // uses Pays::Yes to prevent spam
+        return Err(Error::<T>::UnexpectedCall.into());
+    }
+    fn compute_call_fee_dock_(
+        call: &<T as Config>::Call,
+    ) -> Result<BalanceOf<T>, DispatchErrorWithPostInfo> {
+        use sp_std::convert::TryInto;
+        let fee_fiat_permill: AmountUsd = Self::get_call_fee_fiat_(call)?;
+        let fee_fiat_perbill: u32 = fee_fiat_permill.deconstruct() * 1000;
+
+        let dock_fiat_rate_permill: u32 = Self::dock_fiat_rate().deconstruct();
+        let fee_dock_permill_u64: u64 = (fee_fiat_perbill as u64 * 1_000_u64)
+            .checked_div(dock_fiat_rate_permill as u64)
+            .ok_or(Error::<T>::DivideByZero)?;
+        let fee_dock_permill_u32: u32 = fee_dock_permill_u64
+            .try_into()
+            .or(Err(Error::<T>::ArithmeticOverflow))?;
+
+        // The token has 6 decimal places (defined in the runtime)
+        // pub const DOCK: Balance = 1_000_000;
+        // T::Balance is already expressed in DOCK_Permill or microdock
+        let fee_microdock = <BalanceOf<T>>::from(fee_dock_permill_u32);
+        Ok(fee_microdock)
+    }
+
+    fn charge_fees_(who: T::AccountId, amount: BalanceOf<T>) -> Result<(), DispatchError> {
+        let _ = <T::Currency>::withdraw(
+            &who,
+            amount,
+            WithdrawReasons::FEE,
+            ExistenceRequirement::KeepAlive,
+        )?;
+        Ok(())
+    }
+
+    fn execute_call_(origin: T::Origin, call: &<T as Config>::Call) -> DispatchResultWithPostInfo {
+        let weight = call.get_dispatch_info().weight;
+        // check signature before charging any fees
+        let sender = ensure_signed(origin.clone())?;
+
+        // calculate fee based on type of call
+        let fee_dock = Self::compute_call_fee_dock_(&call)?;
+        // deduct fees based on Currency::Withdraw
+        Self::charge_fees_(sender, fee_dock)?;
+
+        let dispatch_result = call.clone().dispatch_bypass_filter(origin);
+        return match dispatch_result {
+            Ok(_post_dispatch_info) => Ok(PostDispatchInfo {
+                actual_weight: Some(weight),
+                pays_fee: Pays::No,
+            }),
+            Err(e) => {
+                frame_support::debug::info!("{:?}", &e);
+                Err(DispatchErrorWithPostInfo {
```

Comment:
Note: We have an unintentional consequence here: Whether you make a successful or unsuccessful transaction, you pay the same fees. eg. I want to write a DID but it was already written. I still pay the same price as the guy who was able to write the DID. Introducing a "failed-txn" fiat fee helps but its not easy to decide that as you want that to be based on the weight consumed so having a "failed-txn" coefficient that gets multiplied by the weight with the product still less than or equal to successful txn fee seems better.